### PR TITLE
fix: [PL-23148]: Block secret manager deletion if no other SM present

### PR DESCRIPTION
**Pull request by tiwaritejas**
## Describe your changes
Added a check for presence of at least one other secret manager before deleting some secret manager
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39109)

### CODE REVIEW

**Review by shreyas-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39109)* status: 2 file(s) to review. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java` line 482 at r1](https://reviewable.io/reviews/harness/harness-core/39109#-NEuSPo53EmKXc3Cc-82:-NEuSPo53EmKXc3Cc-83:b-e39aox) ([raw file](https://github.com/harness/harness-core/blob/2d9622b59dafff118e512059ca1c6a4646d30cb0/120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java#L482)):*
> ```Java
>     Page<ConnectorResponseDTO> connectorResponseDTOList =
>         list(page, size, accountIdentifier, null, null, null, null, SECRET_MANAGER, null);
>     if (connectorResponseDTOList.getContent().size() < 2) {
> ```

Can we check if size == 1 ?

File: 120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java [hunk 1]
```diff
@@ -467,6 +474,18 @@ public boolean delete(
     }
   }
 
+  private void verifyOtherSecretManagerPresentInAccount(String accountIdentifier, String connectorIdentifier) {
+    int page = 0;
+    int size = 2;
+    Page<ConnectorResponseDTO> connectorResponseDTOList =
+        list(page, size, accountIdentifier, null, null, null, null, SECRET_MANAGER, null);
+    if (connectorResponseDTOList.getContent().size() < 2) {
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java:482]: check if `connectorResponseDTOList.getContent()` will not cause any null pointer

File: 120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java [hunk 2]
```diff
@@ -436,6 +437,12 @@ public boolean delete(
           fullyQualifiedIdentifier, projectIdentifier, orgIdentifier, accountIdentifier, true);
       if (connectorOptional.isPresent()) {
         Connector connector = connectorOptional.get();
+        Boolean isSecretManager =
+            ConnectorRegistryFactory.getConnectorCategory(connector.getType()).equals(SECRET_MANAGER);
+        if (ScopeLevel.of(accountIdentifier, orgIdentifier, projectIdentifier).equals(ScopeLevel.ACCOUNT)
+            && isSecretManager) {
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java:443]: Lets make if condition like
`(isSecretManager && ScopeLevel.of(accountIdentifier, orgIdentifier, projectIdentifier).equals(ScopeLevel.ACCOUNT))` to save a computation

**Review by tiwaritejas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39109)* status: 2 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java` line 443 at r1](https://reviewable.io/reviews/harness/harness-core/39109#-NEuQiyqmb8jjlLQn6VA-r1-443:-NEuaGBZ2dwg-hxXB7Tb:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2d9622b59dafff118e512059ca1c6a4646d30cb0/120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java#L443)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

Lets make if condition like
`(isSecretManager && ScopeLevel.of(accountIdentifier, orgIdentifier, projectIdentifier).equals(ScopeLevel.ACCOUNT))` to save a computation
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java` line 482 at r1](https://reviewable.io/reviews/harness/harness-core/39109#-NEuSPo53EmKXc3Cc-82:-NEuaH3CCdSp6NaatqU3:b-pxd8ny) ([raw file](https://github.com/harness/harness-core/blob/2d9622b59dafff118e512059ca1c6a4646d30cb0/120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java#L482)):*
<details><summary><i>Previously, shreyas-harness (Shreyas Singh) wrote…</i></summary><blockquote>

Can we check if size == 1 ? 
</blockquote></details>

Yes, ideally, we should never be landing in scenario for size == 0. Made the change.
___
*[`120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java` line 482 at r1](https://reviewable.io/reviews/harness/harness-core/39109#-NEuQiyqmb8jjlLQn6VA-r1-482:-NEuaRmn5DemJIM4RutM:b-seqxgn) ([raw file](https://github.com/harness/harness-core/blob/2d9622b59dafff118e512059ca1c6a4646d30cb0/120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java#L482)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

check if `connectorResponseDTOList.getContent()` will not cause any null pointer
</blockquote></details>

I don't think this will occur. We are not making any client calls, and in case there is any issue with DB fetch calls for list connectors, exception will occur before this check.

**Comment by tiwaritejas** [comment]
trigger messagecheck

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39109)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Review by shreyas-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39109)* status: Some of the checks are too old.

**Pull request merged**